posts/simprocs-intro.md

```diff
@@ -0,0 +1,550 @@
+- Try to extract explicit (literal) values for `a` and `b`.
+- Compute `b % a`.
+- If this quantity is zero, then return `True` together with the proof `Nat.dvd_eq_true_of_mod_eq_zero a b rfl`.
+- If this quantity isn't zero, then return `False` together with the proof `Nat.dvd_eq_false_of_mod_ne_zero a b rfl`.
+
+```lean
+builtin_simproc [simp, seval] reduceDvd ((_ : Nat) ∣ _) := fun e => do
+  let_expr Dvd.dvd _ i a b ← e | return .continue
+  let some va ← fromExpr? a | return .continue
+  let some vb ← fromExpr? b | return .continue
+  if vb % va == 0 then
+    return .done { expr := mkConst ``True, proof? := mkApp3 (mkConst ``Nat.dvd_eq_true_of_mod_eq_zero) a b reflBoolTrue}
+  else
+    return .done { expr := mkConst ``False, proof? := mkApp3 (mkConst ``Nat.dvd_eq_false_of_mod_ne_zero) a b reflBoolTrue}
+```
+
+### Many more applications!
+
+At the end of this blog post, we will see how to build step by step a simproc for computing a variant of `List.range` when the parameter is a natural number literal.
+
+## A few caveats
+
+The current design of simprocs comes with a few restrictions that are worth keeping in mind:
+* By definition, **a simproc can only be used in `simp`** (and `simp`-like tactics like `simp_rw`, `simpa`, `aesop`), even though the notion of a "parametric lemma" could be useful in other rewriting tactics like `rw`.
+* **One cannot provide arguments to a simproc to restrict the occurrences it rewrites**.
+  In contrast, this is possible for lemmas in all rewriting tactics: eg `rw [add_comm c]` turns `⊢ a + b = c + d` into `⊢ a + b = d + c` where `rw [add_comm]` would instead have turned it into `⊢ b + a = c + d`.
+* **The syntax for declaring a simproc**, and in particular whether it a simproc should be in the standard simp set or not, **is inconsistent with the rest of the language**: Where we have `lemma` and `@[simp] lemma` to respectively "create a lemma" and "create a lemma and add it to the standard simp set", the analogous constructs for simprocs are `simproc_decl` and `simproc`, instead of `simproc` and `@[simp] simproc`.
+
+## Analogues in other languages
+
+Dsimprocs fill roughly the same niche as proofs by reflection in Rocq.
+
+# How to write a simproc
+
+In this part, we explain how to write a simproc.
+
+First, we detail the inner workings of `simp` relevant to simprocs.
+Then we explain the syntax and general structure of a simproc.
+Finally, we walk through an explicit example of a simproc for a simple custom function.
+
+## How `simp` works
+
+In this section we present some of the inner workings of `simp`.
+
+First we give an overview of the way `simp`, then we delve into the specifics by introducing:
+* The `SimpM` monad, which is the metaprogramming monad holding the information relevant to a `simp` call.
+* `Step`/`DStep`, the Lean representation of a single (definitional) simplification step.
+* `Simproc`/`DSimproc`, the Lean representation of (definitional) simprocs.
```

Comment:
```suggestion
* `Step` (`DStep`), the Lean representation of a single (definitional) simplification step.
* `Simproc` (`DSimproc`), the Lean representation of (definitional) simprocs.
```
or some other version - currently it's not clear what the correspondence actually is. You might also like to explicitly mention that the `D` is for definitional...

Reply:
That's a good point! I think it might be best to keep dsimprocs in the extras, since the discussion is almost identical and having both makes things a little cluttered imo.